### ISSUE TITLE
ikhal & khal: highlight days with events in calendar overview

### ISSUE THREAD

**Issue by smu** [geier/khal#89]
It would be great, if the ikhal calendar overview (left side) could already indicate which days have events. For example these dates could be shown by a bold font.

**Comment by sante81** [comment]
Any develop in this way???

**Comment by libre-man** [comment]
How could this be implemented the best?

Bold text is an option, but that doesn't give any information about the
calendar the event is in. This is especially annoying when there are
calendars that have events for all dates (like I have). So I wouldn't really
like.

We could color the day with the color of the calendar. But what color should
days with multiple events from multiple calenders get? Maybe a priority
setting?

And I don't know what the impact on performance would be, but it shouldn't be
to high.

On Fri, Jun 05, 2015 at 05:38:04AM -0700, Matteo wrote:
> Any develop in this way???
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/geier/khal/issues/89#issuecomment-109281860

-- 
Zorg dat e-mail echt priv� is: onderteken en/of versleutel uw antwoord met 
OpenPGP.

**Comment by untitaker** [comment]
This is sort-of implemented in Android, the color gets black (in our case probably bold) if there are events from multiple calendars. We probably can't do any distinction between all-day events and the rest as Android does.

![](http://www.androidpolice.com/wp-content/uploads/2011/11/wm__0000_2011110214h30_43.jpg)

**Comment by libre-man** [comment]
I think a background color would be better because in the normal khal output
week numbers are bold, so that could cause confusion.

I do think there should be a setting for each calendar if it should mark days
which have an event from that calendar. Some people have weather calendars and
they probably don't want them to mark all the days.

On Fri, Jun 05, 2015 at 05:54:09AM -0700, Markus Unterwaditzer wrote:
> This is sort-of implemented in Android, the color gets black (in our case probably bold) if there are events from multiple calendars. We probably can't do any distinction between all-day events and the rest as Android does.
> 
> ![](http://www.androidpolice.com/wp-content/uploads/2011/11/wm__0000_2011110214h30_43.jpg)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/geier/khal/issues/89#issuecomment-109284069

-- 
Zorg dat e-mail echt privé is: onderteken en/of versleutel uw antwoord met 
OpenPGP.

**Comment by untitaker** [comment]
I wonder if we could use Unicode braille characters to signify the amount of events on one day. This would require at least two chars per day in the calendar walker though.

**Comment by hobarrera** [comment]
I know this is really pick, but I'd the ability to ignore repeating events when highlighting days. Eg:

```
[calendars]

  [[study]]
  path = ~/.local/share/calendars/study/
  color = dark green
  weight: 2
  highlight-repeating: off

  [[work]]
  path = ~/.local/share/calendars/work/
  color = light red
  weight: 2
  highlight-repeating: on
```

The rationale is mostly that I've repeating events almost every day for a single calendar, and it would just pollute the view and make the entire calendar. But I do want to give greater importance to the non-repeating ones.

I'm sure I'm not the only on in this scenario though.

**Comment by geier** [comment]
@untitaker this is only the design of the *old* android calendar, right? Because after some recent update it lost most of its clarity.

**Comment by geier** [comment]
I'm happy to accept pull requests on this :) if anyone wants to work on this, but has  problems with the codebase, I'm happy to supply any hints and/or would make time for a remote pairing session as well.

**Comment by untitaker** [comment]
This is what I am dealing with now, it's similar enough (screenshot attached, hopefully GitHub is smart enough for this)

Note that there's a difference between Google's calendar app and the FOSS one.

**Comment by untitaker** [comment]
Goddamnit.
![_20150605_230423](https://cloud.githubusercontent.com/assets/837573/8015359/226e9fbe-0bd8-11e5-94de-005de75d9b00.JPG)

**Comment by geier** [comment]
ah, I just installed 'Calendar from Android 4.1' and I'm back to the version you have as well. thanks!

**Comment by Keksbringer** [comment]
Is anyone working on this feature yet?
I installed khal today and it was the first thing I missed.
Tried looking through the code for some hours to find out where the default coloring (white on black) takes place to implement at least a workaround that satisfy my needs, but I failed.
Any hints on that? I figured it must be connected to the urwid attribute stuff but all my tries resulted either in nothing or just the wrong behaviour.

**Comment by geier** [comment]
No.

The `ui.calendarwidget.Date`s need to be `urwid.AttrMap`ed with the color you want. If we want to keep the separation we probably need some kind of callback function that gets the color for each Date.

BTW: what color do you envision for Dates with events from more than one calendar?

**Comment by Keksbringer** [comment]
I actually tried changing at least the __init__() of ui.calendarwidget.Date to add some color, but it seems like those colors get overwritten by something else afterwards?
I guess there is also a piece of code that changes the color back after a day was highlighted and that should be changed too?

For multiple calendars I thought about a combination of the color of one calendar (maybe assign priorities in the config file) and using bold numbers.

But for now I was just messing around a bit without spending too much time thinking about a solution for everyone.

**Comment by dzoep** [comment]
Just hacked a support for this at least for khal.

Questions:

1) How do you prefere to receive the patches? github fork and pull request?

2) How would you like to see events from multiple calendars? My suggestion: if only one, use its colour as background, if more than one, use first two colours for the two characters of the date (it at least tells "more than one calendar event here").

3) How should I hook this into configuration? Right now it's hardcoded in calendar/vertical_month/str_week, but it pretty straightforward to make it configurable.

I'm also working on the ikhal integration - probably a hook for the AttrMap of the Date widget should do the trick. Of course I'd love to finish that _after_ it is clear how to display multiple calendars' events and how it should be configurable.

**Comment by untitaker** [comment]
> 1) How do you prefere to receive the patches? github fork and pull request?

Yes.

> 2) How would you like to see events from multiple calendars? My suggestion:
> if only one, use its colour as background, if more than one, use first two
> colours for the two characters of the date (it at least tells "more than one
> calendar event here").

If there is not exactly one calendar color to use, I'd use some default color.
This also helps with calendars that don't have a color.

> 3) How should I hook this into configuration? Right now it's hardcoded in
> calendar/vertical_month/str_week, but it pretty straightforward to make it
> configurable.

Not sure we need to make it configurable.

**Comment by dzoep** [comment]
Ok, you can see it in my fork (dzoep/khal) - right now it uses the colour of the first event found.

@geier - please, what is your opinion about the colouring scheme?

I didn't originally notice it is possible to have calendar without colour defined, that makes things at least ... strange :)

And a side note - why using click and terminal.py? I do promote neither, but I was a bit confused that custom ANSI codes mapping and some library is used together - maybe it would be better to unify it.

**Comment by geier** [comment]
Hi dzoep,
and thanks for taking an interested in khal.

Quoting dzoep (2015-09-30 16:07:16)
> Ok, you can see it in my fork (dzoep/khal) - right now it uses the colour of the first event found.
> 
> @geier - please, what is your opinion about the colouring scheme?

I have really no idea what would be a good idea for days with multiple
colors, perhaps just use bold font? I guess we need to experiment a bit
with this.

> And a side note - why using click and terminal.py? I do promote neither, but I was a bit confused that custom ANSI codes mapping and some library is used together - maybe it would be better to unify it.

urwid uses different color names than click, terminal.py is used for
compatibility with urwid. IIRC urwid supports more colors than click,
i.e. colorama, so I went with urwid's colors.

**Comment by geier** [comment]
Quoting dzoep (2015-09-30 15:35:01)
> 3) How should I hook this into configuration? Right now it's hardcoded in calendar/vertical_month/str_week, but it pretty straightforward to make it configurable.

I'd suggest just adding a new config variable, it will probably be
easiest to just add a new one to [default] and than just pass it
through.

> 
> I'm also working on the ikhal integration - probably a hook for the AttrMap of the Date widget should do the trick. Of course I'd love to finish that _after_ it is clear how to display multiple calendars' events and how it should be configurable.

you might want to have a look at khal.ui.calendarwidget.CListBox._get_color() for this.

**Comment by dzoep** [comment]
Hi,
>
> > And a side note - why using click and terminal.py? I do promote
> neither, but I was a bit confused that custom ANSI codes mapping and
> some library is used together - maybe it would be better to unify it.
>
> urwid uses different color names than click, terminal.py is used for
> compatibility with urwid. IIRC urwid supports more colors than click,
> i.e. colorama, so I went with urwid's colors.
that's probably the best (to use urwid's colors). It allows nice
inclusion of the bold attribute in color name I will stick to that in my
attempts.
> Quoting dzoep (2015-09-30 15:35:01)
> > 3) How should I hook this into configuration? Right now it's
> hardcoded in calendar/vertical_month/str_week, but it pretty
> straightforward to make it configurable.
>
> I'd suggest just adding a new config variable, it will probably be
> easiest to just add a new one to [default] and than just pass it
> through.
>
I had more configuration options on my mind:

Configuration may go like this:

# Enable/disable day coloring (obvious interpretation)
highlight_event_days = 0/1

[highlight_days]

# Highlighting method to use - foreground or background
method = foreground/fg/background/bg

# What color to use when highlighting - explicit color or use calendar color
color = colorname/calendar

# How to color days with events from multiple calendars - either
explicit color or use calendar color
multiple = colorname/calendar

# Default color for calendars without color
default_color = colorname

Of course I am open to any suggestions, but I would like to implement it
completely configurable (as you noted - some experiments are required
and users may have different visual preferences).

> >
> > I'm also working on the ikhal integration - probably a hook for the
> AttrMap of the Date widget should do the trick. Of course I'd love to
> finish that _after_ it is clear how to display multiple calendars'
> events and how it should be configurable.
>
> you might want to have a look at
> khal.ui.calendarwidget.CListBox._get_color() for this.
Thanks for the hint! Will look into it - however I am a bit sceptical
about my idea of multi-colored days...

**Comment by geier** [comment]
Quoting dzoep (2015-09-30 23:12:11)
> I had more configuration options on my mind:
> 
> Configuration may go like this:
> 
> # Enable/disable day coloring (obvious interpretation)
> highlight_event_days = 0/1
> 
> [highlight_days]
> 
> # Highlighting method to use - foreground or background
> method = foreground/fg/background/bg
> 
> # What color to use when highlighting - explicit color or use calendar color
> color = colorname/calendar
> 
> # How to color days with events from multiple calendars - either
> explicit color or use calendar color
> multiple = colorname/calendar
> 
> # Default color for calendars without color
> default_color = colorname
> 
> Of course I am open to any suggestions, but I would like to implement it
> completely configurable (as you noted - some experiments are required
> and users may have different visual preferences).

If the default values are sane, I have no problem with this and am
looking forward to your contribution. If you have any problems, don't
hesitate to contact me.

**Comment by dzoep** [comment]
Ok, I've just finished it all for khal. See https://github.com/dzoep/khal

Do you think it is ready for pull request? I would like to have the configuration considered "stable" before continuing with ikhal.

**Comment by geier** [comment]
It's never too early for a pull request :-) In all seriousness: could you request a pull request? It's much easier to attach comments to a pull request than to the individual commits. I'll have a look at it tonight.

**Comment by geier** [comment]
also: please run `pep8` on your code and fix those pep8 errors (like missing spaces around operators). Thanks!

**Comment by dzoep** [comment]
Cool :)

Created a pull request for the khal part (w/o bold attribute support). I'd love to see some feedback and on saturday plan to work on ikhal part. (And actually... I'm already using it, because it really helps with multiple calendars...).

**Comment by dzoep** [comment]
@geier Would you have some time for online discussion regarding the widgets implementation? Although there's some dead code, I don't wanna create conflict with others' plans. I prefer IRC or XMPP and I am quite flexible time-wise.

**Comment by geier** [comment]
@dzoep my main xmpp account is geier@jabber.ccc.de and I'm cgeier (or cgeier_ ) on freenode.

**Comment by geier** [comment]
closed by #271 and #276 by @dzoep
